importccl: commit planner transaction in planHook

pbardea opened the pull request:
This change commits the planner's transaction during import's plan hook.
It also uses the planner's transaction to create the job record. This
prevents transaction retries from re-running the plan hook and creating
duplicate job records.

Previously, the planner transaction when running the import job may have
retried when the conn executor would auto commit the transaction. Upon
auto commit, uncommited table descriptors would be validated. This
validation could raise a retryable txn error in the case of contention
and cause the entire plan hook to be run again.

Fixes https://github.com/cockroachdb/cockroach/issues/61065.



Release justification: bug fix

Release note (bug fix): Fix a bug where duplicate IMPORT job records
were created due to transaction restarts.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62285)
<!-- Reviewable:end -->

pbardea committed:
importccl: commit planner txn in planHook

Previously, the planner transaction that plans an import statement may
have been auto-retried. This would cause 1 of 2 undesirable behaviours:

1) the retried statement would fail since the first job record was
already running

  - A common failure mode was that when resolving the table name during
planning, it would see the table had been taken OFFLINE by the first job
it created and fail the statement.

2) both executions of the planning code would complete before either got
adopted, and one of the job records would fail when attempting to take
the table OFFLINE since the table descriptor had been modified between
planning and the start of job execution.

Release note (bug fix): Fix a bug where duplicate IMPORT jobs records
may have been created, or IMPORT statements may have failed when the
actual job succeeded.

pbardea commented:
I want to add a bit more commentary about the exact failure modes that we were seeing in the test failures, but it should be ready for an initial review.

dt commented on pkg/ccl/importccl/import_stmt.go:
```diff
@@ -953,6 +953,18 @@ func importPlanHook(
+		if !p.ExtendedEvalContext().TxnImplicit {
+			return errors.AssertionFailedf("expected import planning transaction to be implicit")
+		}
+		// We're about to create the job record. We don't want this transaction
```
More generally: we're about to have some side-effects that are not tied to this txn, which is okay OK if this txn _will_ commit and the only way to be sure of that is to commit it, which we're allowed to do since it is implicit.

pbardea replied:
Updated the comment.

dt commented on pkg/ccl/importccl/import_stmt.go:
```diff
@@ -953,6 +953,18 @@ func importPlanHook(
 
+		// Double check that this transaction is implicit, so that
+		// we're allowed to commit.
+		if !p.ExtendedEvalContext().TxnImplicit {
```
We check this above if not detached, and detached already early returned just above here, so I don't think this is needed?

pbardea replied:
Yep, not needed. It was more of a sanity check just in-case we introduce a regression at the start of the plan hook to ensure that we don't commit an explicit transaction. I'm not super attached to it, and it may have been too defensive so removed it. I updated the comment with a brief explanation instead.

pbardea commented:
TFTR!
bors r=dt
